openwisp_monitoring/db/backends/elasticsearch/index.py

```diff
@@ -0,0 +1,109 @@
+import uuid
+
+from django.conf import settings
+from elasticsearch.exceptions import NotFoundError
+from elasticsearch.helpers import bulk
+from elasticsearch_dsl import Date, Document, InnerDoc, Nested, Q, Search
+from swapper import load_model
+
+# TODO: Remove this. Current Structure
+# Index Templates <-- Indexes <-- Documents <-- Points
+
+
+class Point(InnerDoc):
+    time = Date(required=True, default_timezone=settings.TIME_ZONE)
+    fields = Nested(dynamic=True, required=True, multi=True)
+
+
+class MetricIndex(Document):
+    tags = Nested(dynamic=True, required=False, multi=True)
+    points = Nested(Point)
+
+    class Index:
+        name = 'metric'
+        settings = {
+            'number_of_shards': 1,
+            'number_of_replicas': 0,
+            'lifecycle.name': 'default',
+            'lifecycle.rollover_alias': 'metric',
+        }
+
+
+def find_metric(client, index, tags, retention_policy=None, add=False):
+    search = Search(using=client, index=index)
+    if tags:
+        tags_dict = dict()
+        for key, value in tags.items():
+            tags_dict[f'tags.{key}'] = value
+        q = Q(
+            'nested',
+            path='tags',
+            query=Q(
+                'bool', must=[Q('match', **{k: str(v)}) for k, v in tags_dict.items()]
+            ),
+        )
+    else:
+        q = Q()
+    try:
+        result = list(search.query(q).execute())[0].meta
+        return result['id'], result['index']
+    except (NotFoundError, AttributeError, IndexError):
+        if add:
+            obj = add_doc(client, index, tags, retention_policy=retention_policy)
+            return obj['_id'], obj['_index']
+        return None
+
+
+def add_doc(client, key, tags, _id=None, retention_policy=None):
+    """
+    Add index to elasticsearch using ``keys``, ``tags`` and ``id`` provided.
+    If no ``id`` is provided a random ``uuid`` would be used.
+    """
+    _id = str(_id or uuid.uuid1())
+    # If index exists, add the doc and return
+    try:
+        index_aliases = client.indices.get_alias(index=key)
+        for k, v in index_aliases.items():
+            if v['aliases'][key]['is_write_index']:
+                break
+        client.create(index=k, id=_id, body={'tags': tags})
+        return {'_id': _id, '_index': k}
+    except NotFoundError:
+        pass
+    # Create a new index if it doesn't exist
+    name = f'{key}-000001'
+    obj = MetricIndex(meta={'id': _id})
```

Comment:
`obj`?

Reply:
Basically it's a object of Document class, I can't think of a good name, `obj` is very generic name agreed (maybe that's why they used it in `elasticsearch-dsl docs` :sweat_smile:). Unsure should I name it `doc` or any better suggestion?

Reply:
does `MetricIndex()` generate a new document? A new index? According to the answer to this question, rename the var accordingly